.tekton/pull-request.yaml

```diff
@@ -72,7 +72,7 @@ spec:
                 #!/bin/bash -ex
 
                 python -m pip install gitlint
-                git fetch --unshallow
+                git fetch origin "$(params.target-branch)"
```

Comment:
@amisstea @eisraeli 
this might be enough to address the issue. Please take a look.

Reply:
Wouldn't it be equivalent to `git fetch --update-shallow` ?

Reply:
I agree with @eisraeli, the error suggests a race condition. I think we're just asking for trouble if we continue to modify the local git branches in parallel with any other task like we're still doing here. For a quick fix, has anyone tried setting the `depth` param on the `clone-repository` task to a sufficiently high number yet?

Reply:
I did try changing the depth. gitlint still fails without unshallow in that case (we're missing the history of the remote tracking branch I think).
The error does suggest a race condition, but I suspect the resource being shared between the tasks is the working directory, and with `git fetch origin main` we're not touching the the working directory, so I don't expect any race condition until we have another task that updates that remote-tracking branch.

Having said that, I think we might still need to increase the depth of the clone, but the reason is different. We only have a single commit on this PR so it might not be an issue, but if we have more than one PR then it might be that we need the depth of the PR branch to be at least as big as the number of the commits on the PR.

Reply:
Looking at the debug output from gitlint, I can confirm that it indeed processes more commits from the PR if the depth is increased.
This is with the default depth (1):
```
DEBUG: gitlint.lint Commit Object
--- Commit Message ----
chore: this is going to be deleted

foo bar

Signed-off-by: Yftach Herzog <yherzog@redhat.com>

--- Meta info ---------
Author: Yftach Herzog <yherzog@redhat.com>
Date:   2023-03-15 08:33:01 +0200
is-merge-commit:  False
is-fixup-commit:  False
is-fixup-amend-commit: False
is-squash-commit: False
is-revert-commit: False
Parents: []
Branches: ['(HEAD detached at 315edfc)']
Changed Files: ['.gitignore', '.gitlint', '.tekton/pull-request.yaml', '.yamllint', 'LICENSE', 'README.md', 'automation/promtool_tests.sh', 'foo', 'prometheus/base/kustomization.yaml', 'prometheus/base/namespaces.yaml', 'prometheus/base/prometheus.rules.yaml', 'test/promql/tests/example.yaml', 'test/promql/tests/example2.yaml']
Changed Files Stats:
  .gitignore: 3 additions, 0 deletions
  .gitlint: 10 additions, 0 deletions
  .tekton/pull-request.yaml: 132 additions, 0 deletions
  .yamllint: 26 additions, 0 deletions
  LICENSE: 201 additions, 0 deletions
  README.md: 2 additions, 0 deletions
  automation/promtool_tests.sh: 63 additions, 0 deletions
  foo: 0 additions, 0 deletions
  prometheus/base/kustomization.yaml: 5 additions, 0 deletions
  prometheus/base/namespaces.yaml: 5 additions, 0 deletions
  prometheus/base/prometheus.rules.yaml: 17 additions, 0 deletions
  test/promql/tests/example.yaml: 21 additions, 0 deletions
  test/promql/tests/example2.yaml: 31 additions, 0 deletions
-----------------------
DEBUG: gitlint.cli Exit Code = 0
```

and this is with depth 20:
```
DEBUG: gitlint.lint Commit Object
--- Commit Message ----
chore: this is going to be deleted

foo bar

Signed-off-by: Yftach Herzog <yherzog@redhat.com>

--- Meta info ---------
Author: Yftach Herzog <yherzog@redhat.com>
Date:   2023-03-15 08:33:01 +0200
is-merge-commit:  False
is-fixup-commit:  False
is-fixup-amend-commit: False
is-squash-commit: False
is-revert-commit: False
Parents: ['243033526ce1d7a18b54342e5c44084cccbeedb1']
Branches: ['(HEAD detached at af4ae89)']
Changed Files: ['.tekton/pull-request.yaml', 'foo']
Changed Files Stats:
  .tekton/pull-request.yaml: 3 additions, 1 deletions
  foo: 0 additions, 0 deletions
-----------------------
DEBUG: gitlint.git ('log', '243033526ce1d7a18b54342e5c44084cccbeedb1', '-1', '--pretty=%aN%x00%aE%x00%ai%x00%P%n%B')
DEBUG: gitlint.lint Linting commit 243033526ce1d7a18b54342e5c44084cccbeedb1
DEBUG: gitlint.git ('diff-tree', '--no-commit-id', '--numstat', '-r', '--root', '243033526ce1d7a18b54342e5c44084cccbeedb1')
DEBUG: gitlint.git ('branch', '--contains', '243033526ce1d7a18b54342e5c44084cccbeedb1')
DEBUG: gitlint.lint Commit Object
--- Commit Message ----
fix(STONEO11Y-55): change git fetch command used before gitlint

Using the default depth of 1 required using unshallow later on.
This introduced flakiness as sometimes an error appears for
shallow file has changed since we read it.

With this change, we specifically fetch the target remote branch
which hopefully will work around the flakiness.

Increasing the depth cloning should avoid this flakiness.

Signed-off-by: Yftach Herzog <yherzog@redhat.com>

--- Meta info ---------
Author: Yftach Herzog <yherzog@redhat.com>
Date:   2023-03-13 15:22:43 +0200
is-merge-commit:  False
is-fixup-commit:  False
is-fixup-amend-commit: False
is-squash-commit: False
is-revert-commit: False
Parents: ['1deaca728b3ec6f52563dcb17fad48df5c160728']
Branches: ['(HEAD detached at af4ae89)']
Changed Files: ['.tekton/pull-request.yaml']
Changed Files Stats:
  .tekton/pull-request.yaml: 1 additions, 1 deletions
-----------------------
DEBUG: gitlint.cli Exit Code = 0
```

Notice the additional commit on the second log.

The way I see it, increased depth is required but is not enough. I don't think there's a race condition here if we only fetch the target remote tracking branch.

I suggest going with the approach of fetching the target branch and increasing the depth. If we see the race condition again then we can take more severe actions.

Reply:
@amisstea @eisraeli any thoughts?

Reply:
I understand the replacement of `unshallow` with having a higher `depth` value, but what worries me is actually the `fetch`. I'm not sure if we won't run into the same error of `shallow file has changed since we read it" as I think it relates to multiple tasks modifying the same shallow clone.
Anyway, I'm fine with accepting these changes and be on the lookout for errors resulting from race conditions.

Reply:
There's no other task at the moment that modifies anything in git. I think the problem was that `unshallow` was somehow affected by the changes on the local directory rather than the git metadata. Fetching the remote-tracking branch should not be affected by changes to the local directory.

Reply:
> There's no other task at the moment that modifies anything in git
True.
For now, LGTM.